### PR TITLE
Add DeferredAdapter recursion test

### PR DESCRIPTION
**Pull request by NightlyNexus**
I couldn't find an existing test that demonstrated DeferredAdapter's usage; let me know if I missed one.

### CODE REVIEW

File: moshi/src/test/java/com/squareup/moshi/MoshiTest.java [hunk 1]
```diff
@@ -995,6 +995,39 @@
     }
   }
 
+  @Test public void deferredAdapterRecursesInfinitely() throws Exception {
```
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/MoshiTest.java]: Deferring StackOverflowErrors isn’t really what we’re going for with this feature.

Instead imagine you’re building an adapter for a binary tree:
```
class TreeNode {
  String value;
  TreeNode left;
  TreeNode right;
}
```
The standard reflective adapter for this is going to recursively look itself up, and it’s going to mostly just work.

Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/MoshiTest.java]: And I think we cover this in CircularAdaptersTest. I think those tests will fail if you comment out the deferred adapter code.

Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/MoshiTest.java]: Right.
Sorry, I was reasoning through to myself a few days ago what would happen in this case, as someone was talking to me about not being able to use `nextAdapter` in his factory.
This doesn't make sense as a test.

Comment (NightlyNexus) [inline moshi/src/test/java/com/squareup/moshi/MoshiTest.java]: Oh, I'll give those tests a try. thanks!

**Comment by NightlyNexus** [comment]
👍  CircularAdaptersTest tests.

**Pull request closed without merging**